Custom sliders

ddevault opened the pull request:
Depends on #214, https://github.com/ppy/osu-resources/pull/9, and https://github.com/ppy/osu-framework/pull/384.

Still could use some tweaking but it's a good start, I think.

![](https://sr.ht/GiRR.png)

ddevault committed:
Add sliderbar L+F

ddevault committed:
Play slider sample as the value changes in 1/10ths

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,100 @@
+            };
+        }
+
+        private class OsuSliderBar<U> : SliderBar<U> where U : struct,
```
i realy dislike this formatting. why not move the whole `where U...` to a new line so it's obvious that everything following it is related to U?

peppy commented:
I'm going to have to assume that you don't test anything you code at this rate. Brings back memories of someone else who we had to stop working on the codebase due to the path of destruction they left behind them.

![image](https://cloud.githubusercontent.com/assets/191335/20997322/c08dd4d6-bd47-11e6-9a12-d2757af26add.png)

![](https://puu.sh/sI32y/3df6370b3b.gif)

![image](https://cloud.githubusercontent.com/assets/191335/20997327/cf026e0a-bd47-11e6-972a-5e7fe7191d27.png)

![](https://puu.sh/sI33E/1f20d23210.gif)

border is too thin compared to line
![image](https://cloud.githubusercontent.com/assets/191335/20997342/e9eb9e12-bd47-11e6-8b16-4ec9c8fc5ab6.png)

custom design relies on fucking with base class too much. best off making it not rely on the base framework design through other means (see checkbox implementation maybe).

peppy commented:
if you can't figure how to do this right, there's a completely working and simple implementation in the lazer transitional codebase i wrote. it should be quite easily portable to here

peppy commented:
- sound effect doesn't play

peppy commented:
- left and right arrow key both move right

SirCmpwn commented:
>sound effect doesn't play

osu-resources up to date? works for me

>left and right arrow key both move right

Almost certainly unrelated to my changes, but I'll look into it anyway

>custom design relies on fucking with base class too much. best off making it not rely on the base framework design through other means (see checkbox implementation maybe).

Fair enough

>if you can't figure how to do this right, there's a completely working and simple implementation in the lazer transitional codebase i wrote. it should be quite easily portable to here

I kind of doubt it'll port that easily but I'll have a look

- - -

Regarding the other issues, I knew they were there. Like I said, this is just a start and more work is required. I guess I'll look into fixing them now, though.

ddevault committed:
Fix up slider bugs

ddevault committed:
Move where clause to next line

Tom94 commented:
This is wrong: https://streamable.com/28ar

Also, left and right arrow keys seem to control the wrong sliderbars disregarding which one is hovered: https://streamable.com/np40

Tom94 commented:
Also, I think the sliderbar should perfectly follow the mouse movement, and not lag behind via a delayed transformation. The smoothed movement should only be used with keyboard input.

Tom94 commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,155 @@
-﻿using System;
+﻿using System;
```
Missing license header.

Tom94 commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,155 @@
+            {
+                nub.MoveToX(DrawWidth * value, 300, EasingTypes.OutQuint);
+                leftBox.ScaleTo(new Vector2(
+                        MathHelper.Clamp(DrawWidth * value - nub.Width / 2 + 2, 0, DrawWidth),
```
This clamp is the reason for the first issue I pointed out earlier. I recommend you only apply a transformation to the nub, and inside an overridden `Update` you update the scales of the right and left boxes based on the nub position.

ddevault committed:
Update SliderOption implementation

Per @Tom94's suggestion, the bars track the nub in Update instead of
animating seperately. Also only animates when the event source is the
keyboard.

ddevault committed:
Add license header

SirCmpwn commented:
>Also, left and right arrow keys seem to control the wrong sliderbars disregarding which one is hovered: https://streamable.com/np40

We should talk about this. What's the best solution? I could stash the mouse state and check if it's over our component, but we should probably think about improving focus handling at a broader level (and in seperate PRs).

Other feedback has been addressed.

ddevault committed:
Drop SliderBarEventSource

ddevault committed:
Drop complicated type constraints

peppy commented:
merge master

ddevault committed:
Update submodules

peppy commented:
can't scroll properly. check textbox logic if you're unsure how to fix this.

![](https://puu.sh/sRcZf/e553977db1.gif)

ddevault committed:
Fix transparency issues

peppy commented:
padding wrong

![image](https://cloud.githubusercontent.com/assets/191335/21252798/39d4a5a8-c39d-11e6-947f-e047d6bcf943.png)

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
\ No newline at end of file
+            protected override void UpdateValue(float value)
+            {
+                nub.Position = new Vector2(DrawWidth * value, nub.Position.Y);
```
`nub.MoveToX(DrawWidth * value);`

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+            protected override void Update()
+            {
+                base.Update();
+                leftBox.Scale = new Vector2(MathHelper.Clamp(
```
use DrawWidth instead of DrawSize.X?

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+            protected override bool OnDrag(InputState state)
+            {
+                playSample();
+                return base.OnDrag(state);
```
if we're handling drag here then we `return true`

SirCmpwn replied:
base.OnDrag does so

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+            {
+                if (args.Key == Key.Left || args.Key == Key.Right)
+                    playSample();
+                return base.OnKeyDown(state, args);
```
return true

SirCmpwn replied:
ditto

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+
+            private void playSample()
+            {
+                if (Clock == null)
```
`if (Clock == null || Clock.CurrentTime - lastSample <= 50) return`

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+        private class OsuSliderBar<U> : SliderBar<U> where U : struct
+        {
+            private AudioSample sample;
+            private double lastSample;
```
lastSampleTime

peppy commented on osu.Game/Overlays/Options/SliderOption.cs:
```diff
@@ -1,53 +1,163 @@
+                        RelativeSizeAxes = Axes.None,
+                        Anchor = Anchor.CentreLeft,
+                        Origin = Anchor.CentreLeft,
+                        Colour = new Color4(255, 102, 170, 255),
```
at very leave make the colour a static or member variable. it's used in 3.75 places.

peppy replied:
ColourPalette? sure, go for it. needs to be done ASAP.

SirCmpwn replied:
I do not have all of the colors, but I can work with the colors we're already using in a seperate PR

ddevault committed:
Adjust padding on sliders

ddevault committed:
Update submodules
